### PR TITLE
ISSUE#SB-19247: API to validate html template

### PR DESCRIPTION
**Pull request by aishwa8141**
**Description**: Validates HTML template and returns A flag denoting success

- Downloads the template zip from the given URL in the request

- Unzip the downloaded file

- Checks whether index.html file present in the unzipped folder . If present read the index.html file, fetch all the template variables present in the index.html. Compare with the valid variables.

- If template variables are valid returs flag  valid = true else returns flag valid = false and invalid vars present in template

 

**API endpoint**: cert/v1/template/validate

**Request Format**:

 `{
	"request":{
		"templateUrl":"https://sunbirddev.blob.core.windows.net/user/cert/File-013015823389663232113.zip"
	}
}`

**Response Format** : 

1.  `{
    "id": null,
    "ver": null,
    "ts": null,
    "params": null,
    "responseCode": "OK",
    "result": {
        "response": {
            "valid": true
        }
    }
}`

2. `{
    "id": null,
    "ver": null,
    "ts": null,
    "params": null,
    "responseCode": "OK",
    "result": {
        "response": {
            "valid": false,
            "message": {
                "invalidVars": [
                    "$signatory0Image.jpg"
                ]
            }
        }
    }
}`

3. `{
    "id": null,
    "ver": null,
    "ts": null,
    "params": null,
    "responseCode": "BAD_REQUEST",
    "result": {
        "message": "INVALID_ZIP_FILE :zip file format is invalid, unable to find file index.html"
    }
}`

4. `{
    "id": null,
    "ver": null,
    "ts": null,
    "params": null,
    "responseCode": "BAD_REQUEST",
    "result": {
        "message": "INVALID_ZIP_FILE: exception while unzipping conf/File-013015719310876672678/__MACOSX/css/._style.css (No such file or directory)"
    }
}`

### CODE REVIEW

File: all-actors/src/main/resources/application.conf [hunk 1]
```diff
@@ -117,6 +117,12 @@ certActorSystem {
             nr-of-instances = 5
             dispatcher = cert-dispatcher
           }
+        /TemplateValidateActor
```
Comment (indrajra) [inline all-actors/src/main/resources/application.conf:120]: There are many dispatchers. Pls clean it up and we don't seem to be using anything else other than cert-dispatcher.
The template validation can go to default-dispatcher.

Comment (aishwa8141) [reply all-actors/src/main/resources/application.conf:76]: done

File: all-actors/src/main/java/org/sunbird/cert/actor/TemplateValidateActor.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+
+    private void validateTemplate(Request request) throws BaseException {
+        HTMLValidatorResponse validatorResponse;
+        String templateUrl = (String) request.getRequest().get(JsonKey.TEMPLATE_URL);
+        HTMLTemplateZip htmlTemplateZip = new HTMLTemplateZip(new LocalStore(certsConstant.getDOMAIN_URL()), templateUrl);
+
+        //download the file
+        try {
```
Comment (indrajra) [inline all-actors/src/main/java/org/sunbird/cert/actor/TemplateValidateActor.java:52]: Please try to write less code. Why can't this be a single try-multiple catch blocks - I see invalid zip file or error downloading are the only 2 kinds of errors.
Am finding cleanup line, throw repeated.

File: all-actors/src/main/java/org/sunbird/cert/actor/TemplateValidateActor.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    private HTMLValidatorResponse validateHtml(HTMLTemplateZip htmlTemplateZip) {
+        HTMLValidatorResponse validatorResponse = new HTMLValidatorResponse();
+        String htmlContent;
+        try {
+            htmlContent = htmlTemplateZip.getTemplateContent();
+            HTMLTemplateValidator htmlTemplateValidator = new HTMLTemplateValidator(htmlContent);
+            Set<String> invalidVars = htmlTemplateValidator.validate();
+            if (invalidVars.isEmpty()) {
+                logger.info("template is valid");
```
Comment (indrajra) [inline all-actors/src/main/java/org/sunbird/cert/actor/TemplateValidateActor.java:93]: we must rely on the request id or so to identify which template is valid. Statements like this are pretty much less useful without contextual information. Nothing to do, but letting you recall your other efforts.

File: all-actors/src/main/java/org/sunbird/cert/actor/TemplateValidateActor.java (hunk 2 above)
Comment (aishwa8141) [reply all-actors/src/main/java/org/sunbird/cert/actor/TemplateValidateActor.java:51]: done

**Pull request merged**